raiden/tests/integration/transfer/test_mediatedtransfer.py

```diff
@@ -671,3 +503,150 @@ def get_best_routes_with_fees(*args, **kwargs):
+
+    fee = FeeAmount(5)
+    amount = PaymentAmount(10)
+    cases = [
+        # The fee is added by the initiator, but no mediator deducts fees. As a
+        # result, the target receives the fee.
+        dict(
+            fee_schedules=[None, None, None],
+            expected_transferred_amounts=[amount + fee, amount + fee, amount + fee],
+        ),
+        # The first mediator claims all of the fee.
+        dict(
+            fee_schedules=[None, FeeScheduleState(flat=fee), None],
+            expected_transferred_amounts=[amount + fee, amount, amount],
+        ),
+        # The first mediator has a proportional fee of 20%
+        dict(
+            fee_schedules=[None, FeeScheduleState(proportional=0.20e6), None],
+            expected_transferred_amounts=[
+                amount + fee,
+                amount + fee - (amount + fee) // 5,
+                amount + fee - (amount + fee) // 5,
+            ],
+        ),
+        # Both mediators have a proportional fee of 20%
+        dict(
+            fee_schedules=[
+                None,
+                FeeScheduleState(proportional=0.20e6),
+                FeeScheduleState(proportional=0.20e6),
+            ],
+            expected_transferred_amounts=[
+                amount + fee,
+                amount + fee - (amount + fee) // 5,
+                amount + fee - (amount + fee) // 5 - (amount + fee - (amount + fee) // 5) // 5,
+            ],
+        ),
+        # The first mediator has an imbalance fee that works like a 20%
+        # proportional fee when using the channel in this direction.
+        dict(
+            fee_schedules=[None, FeeScheduleState(imbalance_penalty=[(0, 0), (1000, 200)]), None],
+            expected_transferred_amounts=[
+                amount + fee,
+                amount + fee - (amount + fee) // 5,
+                amount + fee - (amount + fee) // 5,
+            ],
+        ),
+        # Using the same fee_schedules as above on the incoming channel instead
+        # of the outgoing channel of mediator 1 should yield the same result.
+        dict(
+            incoming_fee_schedules=[
+                FeeScheduleState(imbalance_penalty=[(0, 0), (1000, 200)]),
+                None,
+                None,
+            ],
+            expected_transferred_amounts=[
+                amount + fee,
+                amount + fee - (amount + fee) // 5,
+                amount + fee - (amount + fee) // 5,
+            ],
+        ),
+        # The first mediator has an imbalance fee which will add one token for
+        # for every token transferred as a reward for moving the channel into a
+        # better state. This causes the target to receive more than the `amount
+        # + fees` which is sent by the initiator.
+        dict(
+            fee_schedules=[None, FeeScheduleState(imbalance_penalty=[(0, 1000), (1000, 0)]), None],
+            expected_transferred_amounts=[amount + fee, (amount + fee) * 2, (amount + fee) * 2],
+        ),
```

Comment:
The cases look nice. I wonder if we should add something like case 2, only with `flat_fee = fee + 1`. Then nothing should be transferred.

Reply:
In its current form, this only works for happy cases tests. Otherwise we run into timeouts. This is the same reason why I didn't merge `test_mediated_transfer_with_node_consuming_more_than_allocated_fee` with this test, which mostly covers what you request.